src/Groups/matrices/iso_oscar_gap.jl

```diff
@@ -123,30 +123,29 @@ end
 #
 #  Matrix space isomorphism
 #
+#  Using the known ring isomorphism from an Oscar ring to a GAP ring,
+#  we can map matrices from Oscar to GAP using `map_entries`.
+#  (The generic `map_entries` method cannot be used because the concepts of
+#  `parent`and `_change_base_ring` do not fit to the situation in GAP.)
+#  For the direction from GAP to Oscar, we introduce a generic function
+#  `preimage_matrix` that takes the `ring_iso` and a GAP matrix.
+#
 ################################################################################
 
-# computes the isomorphism between the Oscar matrix space of dimension deg over
-# F and the corresponding GAP matrix space
-
-mat_iso_oscar_gap(F::Union{Nemo.GaloisField, Nemo.GaloisFmpzField, Nemo.FqNmodFiniteField, Nemo.FqFiniteField, Nemo.FlintRationalField, Nemo.AnticNumberField}, deg) = mat_iso_oscar_gap(F, deg, ring_iso_oscar_gap(F))
-
-function mat_iso_oscar_gap(F::T, deg::Int, FtoGAP::MapFromFunc{T, S}) where {T <: Union{Nemo.GaloisField, Nemo.GaloisFmpzField, Nemo.FqNmodFiniteField, Nemo.FqFiniteField, Nemo.FlintRationalField, Nemo.AnticNumberField}, S}
-   function f(x::MatElem)
-      @assert base_ring(x) === domain(FtoGAP)
-      rows = Vector{GapObj}(undef, nrows(x))
-      for i in 1:nrows(x)
-         rows[i] = GapObj([ FtoGAP(x[i, j]) for j in 1:ncols(x) ])
-      end
-      return GAP.Globals.ImmutableMatrix(codomain(FtoGAP), GapObj(rows), true)
-   end
-
-   function finv(x::GapObj)
-      m = GAP.Globals.NrRows(x)
-      n = GAP.Globals.NrCols(x)
-      L = [ preimage(FtoGAP, x[i, j]) for i in 1:m for j in 1:n]
-      return matrix(domain(FtoGAP), m, n, L)
+function AbstractAlgebra.map_entries(f::Map{T, GapObj}, a::MatElem) where T
+   isempty(a) && error("empty matrices are not supported by GAP")
+   @assert base_ring(a) === domain(f)
+   rows = Vector{GapObj}(undef, nrows(a))
+   for i in 1:nrows(a)
+      rows[i] = GapObj([f(a[i, j]) for j in 1:ncols(a)])
    end
+   return GAP.Globals.ImmutableMatrix(codomain(f), GapObj(rows), true)
+end
 
-   M = MatrixSpace(F, deg, deg)
-   return MapFromFunc(f, finv, M, GAP.Globals.MatrixAlgebra(codomain(FtoGAP), deg))
+function preimage_matrix(f::Map{T, GapObj}, a::GapObj) where T
```

Comment:
depends on how much gap-matrices should support the matrix interface? Or possibly, how to interact with the non-types gap-objects as the mat-interface possibly makes no sens for gap-objects representing groups (or group elems). I guess s.th. that uses map pr map_entries paired with an additional argument might work.
Or, a thin new object GapMat containing the gap-object and implementing the neccessary matrix functions. Then map_entries(f, GapMat(g)) could work fine?

Reply:
Perhaps I am missing the point.
We could install the following `map_entries` method instead of `preimage_matrix`.
```
function AbstractAlgebra.map_entries(f::Map{GAP.GapObj, T}, a::GAP.GapObj) where T
   m = GAP.Wrappers.NumberRows(a)
   n = GAP.Wrappers.NumberColumns(a)
   L = [f(a[i, j]) for i in 1:m for j in 1:n]
   return matrix(codomain(f), m, n, L)
end
```
Then `map_entries(finv, GAP.evalstr("[[Z(3), Z(3)^0],[0*Z(3), Z(3)]]"))` would work,
where `finv = inv(f)` for `f = ring_iso_oscar_gap(GF(3))`. (For that, one needs thofma/Hecke.jl/pull/447.)

The point is that `f` is stored in the matrix group or form object, but in order to create Oscar matrices from GAP matrices via `map_entries`, we would have to call `inv(f)` (and hence create a new object, since the result is not cached) again and again, or we could store `finv` in the matrix group or form object (also not attractive, we are happy that we can get rid of the `mat_iso` field, we do not want to introduce a new field `inv_ring_iso` or so).

My `preimage_matrix` is nothing but `map_entries` for the inverse map.